usr/lib/common/mech_openssl.c

```diff
@@ -0,0 +1,4617 @@
+/*
+ * COPYRIGHT (c) International Business Machines Corp. 2021
+ *
+ * This program is provided under the terms of the Common Public License,
+ * version 1.0 (CPL-1.0). Any use, reproduction or distribution for this
+ * software constitutes recipient's acceptance of CPL-1.0 terms which can be
+ * found in the file LICENSE file or at
+ * https://opensource.org/licenses/cpl1.0.php
+ */
+
+
+#include <pthread.h>
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+
+#include <openssl/opensslv.h>
+
+#include "pkcs11types.h"
+#include "defs.h"
+#include "host_defs.h"
+#include "h_extern.h"
+#include "tok_spec_struct.h"
+#include "trace.h"
+
+#include <openssl/crypto.h>
+#include <openssl/err.h>
+#include <openssl/evp.h>
+#include <openssl/rsa.h>
+#include <openssl/ec.h>
+#include <openssl/bn.h>
+#include <openssl/sha.h>
+#include <openssl/cmac.h>
+#include <openssl/des.h>
+#if OPENSSL_VERSION_PREREQ(3, 0)
+#include <openssl/core_names.h>
+#include <openssl/param_build.h>
+#endif
+
+CK_RV openssl_specific_rsa_keygen(TEMPLATE *publ_tmpl, TEMPLATE *priv_tmpl)
+{
+    CK_ATTRIBUTE *publ_exp = NULL;
+    CK_ATTRIBUTE *attr = NULL;
+    CK_ULONG mod_bits;
+    CK_BBOOL flag;
+    CK_RV rc;
+    CK_ULONG BNLength;
+#if !OPENSSL_VERSION_PREREQ(3, 0)
+    const RSA *rsa = NULL;
+    const BIGNUM *bignum = NULL;
+#else
+    BIGNUM *bignum = NULL;
+#endif
+    CK_BYTE *ssl_ptr = NULL;
+    BIGNUM *e = NULL;
+    EVP_PKEY *pkey = NULL;
+    EVP_PKEY_CTX *ctx = NULL;
+
+    rc = template_attribute_get_ulong(publ_tmpl, CKA_MODULUS_BITS, &mod_bits);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("%s\n", ock_err(ERR_TEMPLATE_INCOMPLETE));
+        return CKR_TEMPLATE_INCOMPLETE; // should never happen
+    }
+
+    // we don't support less than 512 bit keys in the sw
+    if (mod_bits < 512 || mod_bits > 4096) {
```

Comment:
The soft token itself still uses 4096 as max key size supported in the mechanisms info. I could actually use OPENSSL_RSA_MAX_MODULUS_BITS for the max key size for the RSA mechs in the soft token, too. If that's wanted, then I guess this would better be a separate/new commit. 
Opinions?